core/src/main/scala/org/apache/spark/TaskContextImpl.scala

```diff
@@ -82,9 +97,27 @@ private[spark] class TaskContextImpl(
     }
   }
 
-  /** Marks the task for interruption, i.e. cancellation. */
+  /**
+   * Marks the task as interruption, i.e. cancellation. We add this
+   * method for some more clean works. For example, we need to register
+   * an "interruption" callback to completely stop a receiver supervisor.
+   */
   private[spark] def markInterrupted(): Unit = {
     interrupted = true
+    val errorMsgs = new ArrayBuffer[String](2)
+    // Process interruption callbacks in the reverse order of registration
+    onInterruptedCallbacks.reverse.foreach { listener =>
+      try {
+        listener.onTaskInterrupted(this)
+      } catch {
+        case e: Throwable =>
```

Comment:
It might be dangerous for this code to catch all Throwables, since this could also catch irrecoverable exceptions like OutOfMemoryError.  Instead, it might be safer to use `case NonFatal(e) =>`.  If you agree, then let's also update the original code in `markTaskCompleted()` to catch NonFatal as well.